webpack/scenes/ModuleStreams/__tests__/ModuleStreamPage.test.js

```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import { shallow } from 'enzyme';
+import toJson from 'enzyme-to-json';
+import ModuleStreamsPage from '../ModuleStreamsPage';
+import ModuleStreamsTable from '../ModuleStreamsTable';
+import Search from '../../../components/Search/index';
+
+jest.mock('../../../move_to_foreman/foreman_toast_notifications');
+
+describe('Module streams page', () => {
+  it('should render and contain appropiate components', async () => {
+    const moduleStreams = {};
+    const mockLocation = { search: '' };
+    const getModuleStreams = () => {};
+
+    const wrapper = shallow(<ModuleStreamsPage
+      moduleStreams={moduleStreams}
+      getModuleStreams={getModuleStreams}
+      location={mockLocation}
+    />);
+
+    expect(toJson(wrapper)).toMatchSnapshot();
+    expect(wrapper.find(ModuleStreamsTable)).toHaveLength(1);
+    expect(wrapper.find(Search)).toHaveLength(1);
```

Comment:
Isn't the snapshot enough in this case?

Reply:
Why not test the expected child components are rendered? I suppose the snapshot will test that, but snapshots are easily re-recorded and these explicit tests make the main purpose of the component be more of a constant.

I'm also trying to leverage jest/enzyme for more than just snapshot testing, they are powerful libraries that can do a lot, this is something I think we should start to use more

Reply:
IMHO snapshots should not be easily re-recorded and I personally always checking them in reviews.
I believe it would be much easier to maintain tests that looks similar to each other that only produce snapshots.

In foreman we use a test helper so our tests are more data-driven:
https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/BreadcrumbBar/components/Breadcrumb.test.js

Reply:
I mean they are easily re-recorded by nature of being a snapshot and using a simple command to re-record them. I also check them when adding new code and in PRs (I definitely wasn't suggesting to not too), but there is no guarantee that others will too, espeically those new to React. I prefer to make explicit statements about what is expected in the rendered component since updating these tests is more deliberate.

My larger concern is that we rely on snapshots too much, they are just one part of testing and there are tools we could use to do more testing on the front-end, but I think that is a discussion outside the scope of this PR :)

Its good to know about the test helpers, I will have to look at them more. Are you suggesting I use them in this PR or were you showing them for general knowledge?

Reply:
I don't think you should rewrite your tests (unless you have time and you want to) but I would love to see it getting used more as a standard.
I brought it to the katello repo until we can find a solution for consuming it (from foreman/patternfy/separate-npm-package):
https://github.com/Katello/katello/blob/master/webpack/move_to_pf/test-utils/testHelpers.js

In this PR I refactored some parts of the subscriptions-page and uses the test-helpers when testing redux actions/reducers/selectors:
https://github.com/Katello/katello/pull/7583/files#diff-39739310d70dd9189f1f39cae7c44058
https://github.com/Katello/katello/pull/7583/files#diff-f0e6881d0de88b03101a4a004159c14b
https://github.com/Katello/katello/pull/7583/files#diff-5cdd0d73044c96085317cd727988324f
https://github.com/Katello/katello/pull/7583/files#diff-737a86892912d238ef5ddcfc4bd09183
https://github.com/Katello/katello/pull/7583/files#diff-f1df5a73c63d0fc657fc231d6cf37af8
https://github.com/Katello/katello/pull/7583/files#diff-24c6485c3b06c73a6d832c4170bf999d
https://github.com/Katello/katello/pull/7583/files#diff-2102bc95e8398f0df9d4847dc0a13141


I gave some example but you can notice I am only use snapshots and trying to embrace it over manual code.
I love them so much because:
1. Instead telling me there is an error, it shows me exactly the problem and what went wrong.
2. It's easier to write those tests and you don't need to think to much about logic.
3. It covers much more than manual code that testing something very specific.
4. Usually you shouldn't update the test logic, only the "expected-results" which moved to a separate file we can track changes in over time.

I think that when it comes to new developers/react-developers, tests feel like something they don't like and fear of dealing with. We should help them and educate them about tests and try to make the tests feel more comfortable for them. I think snapshots and generic tests help for achieving it.

Reply:
Thanks for sharing the examples, Those tests look really good and are definitely are in the direction I was thinking as well. I think that is a great example of using snapshots to test multiple types of functionality.

I agree that testing can be daunting for new react developers, this may be a good candidate for a deep dive! :mag: